File: 2555.maximize-win-from-two-segments.py
#
# @lc app=leetcode id=2555 lang=python3
#
# [2555] Maximize Win From Two Segments
#

# @lc code=start
from collections import Counter
class Solution:
    def maximizeWin(self, prizePositions: list[int], k: int) -> int:
        # n = len(prizePositions)
        # ans = left = 0
        # mx = [0] * (n + 1)
        # for right, p in enumerate(prizePositions):
        #     while p - prizePositions[left] > k:
        #         left += 1
        #     ans = max(ans, mx[left] + right - left + 1)
        #     mx[right + 1] = max(mx[right], right - left + 1)
        # return ans
        



        #todo redo it again to see if I improve my skills or not
        n = len(prizePositions)
        ans  = left = 0
        mx_num = [0] * (n + 1)
        for right, p in enumerate(prizePositions):
            while p - prizePositions[left] > k:
                left += 1
            ans = max(ans, mx_num[left] + right - left + 1)
            mx_num[right + 1] = max(mx_num[right], right - left + 1)
        return ans
# @lc code=end

File: 2555.maximize-win-from-two-segments.java
/*
 * @lc app=leetcode id=2555 lang=java
 *
 * [2555] Maximize Win From Two Segments
 */

// @lc code=start
class Solution {
    public int maximizeWin(int[] prizePositions, int k) {
        int ans = 0, n = prizePositions.length, left = 0;
        int[] mx = new int[n + 1 ];
        if (2 * k + 1 >= prizePositions[n - 1] - prizePositions[0]) {
            return n;
        }
        for(int right = 0; right < n; right++){
            while(prizePositions[right] - prizePositions[left] > k){
                left++;
            }
            ans = Math.max(ans, mx[left] + right - left + 1);
            mx[right + 1] = Math.max(mx[right], right - left + 1);
        }
        return ans; 
    }
}
// @lc code=end

